index.bs

```diff
@@ -1715,3 +1705,220 @@ When [=a browsing context is discarded=], run the following steps:
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+        GetRealmsResult = {
+          realms: [*RealmInfo]
+        }
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for script.getRealms">
+The [=remote end steps=] with |command parameters| are:
+
+  1. Let |environment settings| be a list of all the [=environment settings objects=]
+     that have their [=execution ready flag=] set.
+
+  1. If |command parameters| contains <code>context</code>:
+
+    1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+       with |command parameters|["<code>context</code>"].
+
+    1. Let |document| be |context|'s [=active document=].
+
+    1. Let |context environment settings| be a list.
+
+    1. For each |settings| of |environment settings|:
+
+      1. If any of the following conditions hold:
+
+        * The [=responsible document=] of |settings| is |document|
+
+        * The [=Realm/global object=] specified by |settings| is a
+          {{WorkerGlobalScope}} with |document| in its [=owner set=]
+
+        Append |settings| to |context environment settings|.
+
+    1. Set |environment settings| to |context environment settings|.
+
+  1. Let |realms| be a list.
+
+  1. For each |settings| of |environment settings|:
+
+    1. Let |realm info| be the result of [=get the realm info=] given |settings|
+
+    1. If |realm info| is not null, append |realm info| to |realms|.
+
+  1. Let |body| be a map matching the <code>GetRealmsResult</code> production,
+     with the <code>realms</code> field set to |realms|.
+
+  1. Return [=success=] with data |body|.
+
+
+Issue: Extend this to also allow realm parents e.g. for nested workers? Or get all ancestor workers.
+
+</div>
+
+### Events ### {#module-script-events}
+
+#### realmCreated #### {#event-realm-created}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        RealmCreatedEvent = {
+         method: "script.realmCreated",
+         params: RealmInfo
+       }
+      </pre>
+   </dd>
+</dl>
+The [=remote end event trigger=] is:
+
+<div algorithm="remote end event trigger for script.realmCreated"> When any
+of the [=set up a window environment settings object=], [=set up a worker
+environment settings object=] or [=set up a worklet environment settings
+object=] algorithms are invoked, immediately prior to returning the settings
+object:
+
+ 1. Let |environment settings| be the newly created [=environment settings
+    object=].
+
+ 1. Let |realm info| be be the result of [=get the realm info=] given
+    |environment settings|.
+
+ 1. Let |related contexts| be an empty set.
+
+ 1. If the [=responsible document=] of |settings| is a [=Document=], append the
+    [=responsible document=]'s [=Document/browsing context=] to |related contexts|.
+
+    Otherwise if the [=Realm/global object=] specified by |settings| is a
+    {{WorkerGlobalScope}}, for each |owner| in the [=Realm/global object=]'s [=owner
```

Comment:
OK, so the "for each" here is how we can end up with multiple members in *related contexts*, which can happen for shared workers.

However, a shared worker is only created once right, and some browsing context has to be the one that brought it about. Can there in fact be multiple owners at the time a realm is created?

In the case that a browsing context later joins the party with a shared worker, is an event needed to signal that, or do we pretent it was created again?

Reply:
I think we could at least have multiple contexts associated with a worker if we are emitting events because we subscribed to the `script.realmCreated` event for some existing contexts and we're backfilling. The spec doesn't cover that yet but I think it will.

I don't know what CDP does with shared workers; could be interesting to investigate.

Reply:
OK so let's not block this PR on investigating that, seems like a fine thing to look at while prototyping an implementation.